In [2]:
# Import required libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LinearRegression

from sklearn.metrics import (
    mean_absolute_error,
    mean_squared_error,
    r2_score
)

In [3]:
# Load the dataset

df = pd.read_csv("insurance.csv")

# Task 1: Data Understanding

In this task, we load the dataset, inspect its structure, display the first five records, and identify the numerical features, categorical features, and the target variable.

In [5]:
# Display the first five records

df.head()

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520


In [7]:
# Display information about the dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB


In [8]:
# Number of rows and columns

print("Dataset Shape:", df.shape)

Dataset Shape: (1338, 7)


In [9]:
# Display all column names

print("Columns in Dataset:")
print(df.columns.tolist())

Columns in Dataset:
['age', 'sex', 'bmi', 'children', 'smoker', 'region', 'charges']


In [10]:
# Statistical summary of numerical columns

df.describe()

,age,bmi,children,charges
count,1338.000000,1338.000000,1338.000000,1338.000000
mean,39.207025,30.663397,1.094918,13270.422265
std,14.049960,6.098187,1.205493,12110.011237
min,18.000000,15.960000,0.000000,1121.873900
25%,27.000000,26.296250,0.000000,4740.287150
50%,39.000000,30.400000,1.000000,9382.033000
75%,51.000000,34.693750,2.000000,16639.912515
max,64.000000,53.130000,5.000000,63770.428010


In [11]:
# Identify numerical features

numerical_features = df.select_dtypes(include=["int64", "float64"]).columns.tolist()

print("Numerical Features:")
print(numerical_features)

Numerical Features:
['age', 'bmi', 'children', 'charges']


In [12]:
# Identify categorical features

categorical_features = df.select_dtypes(include=["object"]).columns.tolist()

print("Categorical Features:")
print(categorical_features)

Categorical Features:
['sex', 'smoker', 'region']


In [13]:
# Define the target variable

target_variable = "charges"

print("Target Variable:")
print(target_variable)

Target Variable:
charges


## Observations

- The dataset contains **1338 records** and **7 columns**.
- There are **4 numerical features**: age, bmi, children, and charges.
- There are **3 categorical features**: sex, smoker, and region.
- The target variable is **charges**, which represents the medical insurance cost.

## Observations

- The dataset contains **1338 records** and **7 columns**.
- There are **4 numerical features**: age, bmi, children, and charges.
- There are **3 categorical features**: sex, smoker, and region.
- The target variable is **charges**, which represents the medical insurance cost.

In [14]:
# Check for missing values in each column

missing_values = df.isnull().sum()

print("Missing Values:")
print(missing_values)

Missing Values:
age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64


In [15]:
# Create a LabelEncoder object

label_encoder = LabelEncoder()

In [19]:
# Encode 'sex'
df["sex"] = label_encoder.fit_transform(df["sex"])

# Encode 'smoker'
df["smoker"] = label_encoder.fit_transform(df["smoker"])

# Encode 'region'
df["region"] = label_encoder.fit_transform(df["region"])

In [ ]:
# Display the first five records after encoding

df.head()

,age,sex,bmi,children,smoker,region,charges
0,19,0,27.900,0,1,3,16884.92400
1,18,1,33.770,1,0,2,1725.55230
2,28,1,33.000,3,0,2,4449.46200
3,33,1,22.705,0,0,1,21984.47061
4,32,1,28.880,0,0,1,3866.85520


In [23]:
# Separate input features (X) and target variable (y)

X = df.drop("charges", axis=1)

y = df["charges"]

In [24]:
print("Features Shape:", X.shape)
print("Target Shape:", y.shape)

Features Shape: (1338, 6)
Target Shape: (1338,)


In [25]:
# Split the dataset into training and testing sets

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42
)

In [26]:
print("Training Features Shape:", X_train.shape)
print("Testing Features Shape:", X_test.shape)

print("Training Target Shape:", y_train.shape)
print("Testing Target Shape:", y_test.shape)

Training Features Shape: (1070, 6)
Testing Features Shape: (268, 6)
Training Target Shape: (1070,)
Testing Target Shape: (268,)


## Observations

- The dataset contains no missing values, so no data cleaning was required.
- The categorical features (`sex`, `smoker`, and `region`) were successfully encoded into numerical values using LabelEncoder.
- The dataset was split into **80% training data (1070 records)** and **20% testing data (268 records)** for model training and evaluation.